Admin pages: Assignments Table, Assignment Profile, Unassigned

LWest001 opened the pull request:
## Sets up the Admin view Assignments table. 

![image](https://github.com/codeforboston/urban-league-heat-pump-accelerator/assets/93659375/1116e088-d6de-49a4-a1f0-1763d75673f2)


Restructured table a bit:
- 'Surveyor' column changed to 'Surveyor(s)' and should display links to as many surveyors are put on an assignment.
- Removed 'Home' column - function of this column unknown as one Assignment may have many homes
- Removed 'Surveyed' column - function of this column unknown.
- 'Completed' column changed to 'Completion' and displays a count of (completed homes / total homes in assignment). Uses a non-exist 'completed' property from each home object within an assignment. This will probably have to be worked out on the backend. Clarification please on what a "completed" assignment/home is.
- 'Assignment' column provides links to assignments

## Sets up Admin view Assignment pages for single assignments (clicking 'View' under 'Assignment' column on the Assignments table navigates here).
![image](https://github.com/codeforboston/urban-league-heat-pump-accelerator/assets/93659375/73de122f-31fb-4db0-8803-05d0d6b054f6)

- Lists homes data inside of an assignment cluster.
- No access yet to unassign homes.
- Also makes use of the non-existant 'completed' property.

thiagobardini commented:
My use case suggestions: @LWest001 @Rando-cal @mattdelsordo 

### Assignments page - URL: /admin/assignment

**Adding and Removing Surveyors from Assignments**

1. Select the rows you want to assign to the Assignments by clicking on the checkbox.
1. Open the Surveyor dropdown located at the top.
    - If surveyors have already been added, ensure that the appropriate surveyors are selected.
1. Click on the surveyors you wish to add or remove from the selected Assignments.
   - Note: A maximum of two surveyors can be assigned per Assignment.

**Maximum Number of Surveys per Assignment = 23**
The Google Maps API restricts the display of more than 23 houses per route.

**Completion Column**
When you complete all the surveys assigned to an Assignment, the Assignment is marked as completed, as shown in the example 23/23.

**Regarding completion**
When you finish a survey and answer all the questions, mark the survey as complete.

### Assignment ID - URL: /admin/assignment/assignProfile/1

**Unassign Surveys/Homes from the Assignment**
1. Click on the Assignment button (URL: /admin/assignment).
    - Redirect to the URL: /admin/assignment/assignProfile/1.
1. Click on the Unassign X button.
   - This action automatically moves the Home/Survey related to the Assignment to the Unassignment page (URL: /admin/assignment/unassigned).

### Unassigned Page - URL: /admin/assignment/unassigned

**Create a New Assignment**
At the top, we need a "Create Assignment" button.

1. Click on the checkbox in the row table.
    - Maximum number of surveys per Assignment: 23.
2. After selecting the checkbox rows (homes/surveys), click on the "Create Assignment" button to create an Assignment ID.

**Add Homes/Surveys to Existing Assignment ID**
1. Click on the Assignment dropdown at the top where you want to assign more Homes/Surveys.
   - After selecting from the dropdown, verify that the Assignment ID number is visible.
2. In the table, click on the checkboxes of the Homes/Surveys you want to assign to the Assignment ID.
    - Note: The maximum number of homes/surveys per assignment is 23.
3. Click on the "Send To" button to finalize the assignment.
    - After clicking, verify that all the Homes/Surveys are no longer shown.
    - Verify that the added Homes/Surveys are displayed under the Assignment ID on the URL: /admin/assignment.


### Let me know if you have any suggestions or if we can move forward with it. Thanks!

LWest001 commented:
> ### Assignments page - URL: /admin/assignment
> 1. Select the rows you want to assign to the Assignments by clicking on the checkbox.
> 2. Open the Surveyor dropdown located at the top.
>    * If surveyors have already been added, ensure that the appropriate surveyors are selected.
> 3. Click on the surveyors you wish to add or remove from the selected Assignments.
>    
>    * Note: A maximum of two surveyors can be assigned per Assignment.

I fixed the dropdown menu in the newest set of commits.

> **Maximum Number of Surveys/Homes per Assignment = 23** The Google Maps API restricts the display of more than 23 houses per route.

Good note. I think this is something we should implement on the Homes page, since that is where homes can be added to assignments

> **Completion Column** When you complete all the surveys assigned to an Assignment, the Assignment is marked as completed, as shown in the example 23/23.
>
> **Regarding completion** When you finish a survey and answer all the questions, mark the survey as complete.

Agreed. To implement this on request might be kind of expensive. We would have to check all `survey_visits`, check their `survey_response` to see if all their `survey_answers` are filled, check the `home` that the `survey_visit` belongs to, then see which `assignment` the `home` belongs to... 
I think this should be done on the backend so we can just get a `completed` property attached to a `home`, then check only which homes belong to which assignments and if they are completed or not.

> ### Assignment ID - URL: /admin/assignment/assignProfile/1
> **Unassign Surveys/Homes from the Assignment**
> 
> 1. Click on the View button on the Assignment column (URL: /admin/assignment).
>    
>    * Redirect to the URL: /admin/assignment/assignProfile/1.
> 2. Click on the Unassign X button.
>    
>    * This action automatically moves the Home/Survey related to the Assignment to the Unassignment page (URL: /admin/assignment/unassigned).

Yes - I think this function should be implemented in a separate PR. The scope of this one should be querying and displaying data and visual tweaks. We have a whole batch of mutation-related issues to tackle!

> ### Unassigned Page - URL: /admin/assignment/unassigned

I went ahead and implemented data loading and presentation in the newest batch of commits for both unassigned homes (via an apiSlice hook) and a list of assignments in the dropdown menu.

> **Create a New Assignment** At the top, we need a "Create Assignment" button.
> 
> 1. Click on the checkbox in the row table.
>    
>    * Maximum number of surveys per Assignment: 23.
> 2. After selecting the checkbox rows (homes/surveys), click on the "Create Assignment" button to create an Assignment ID.

Definitely need this function - I implemented it as a "Create New" option in the dropdown. Let me know what you think.

> **Add Homes/Surveys to Existing Assignment ID**
> 
> 1. Click on the Assignment dropdown at the top where you want to assign more Homes/Surveys.
>    
>    * After selecting from the dropdown, verify that the Assignment ID number is visible.
> 2. In the table, click on the checkboxes of the Homes/Surveys you want to assign to the Assignment ID.
>    
>    * Note: The maximum number of homes/surveys per assignment is 23.
> 3. Click on the "Send To" button to finalize the assignment.
>    
>    * After clicking, verify that all the Homes/Surveys are no longer shown.
>    * Verify that the added Homes/Surveys are displayed under the Assignment ID on the URL: /admin/assignment.

Yes, let's make sure these parameters are in place when we implement the mutation requests!

> ### Let me know if you have any suggestions or if we can move forward with it. Thanks!

One thing I would like to see is a user verification before adding homes to an assignment. A good way to do this would be a modal that pops up when the Add Homes To Assignment button is clicked. The Modal would list the homes on that assignment, and maybe show where that home relates to a map of the existing homes in the assignment. Then, the user would click either "Add Home To This Assignment" or "Cancel".

mattdelsordo commented on frontend/front/src/api/apiSlice.js:
```diff
@@ -29,6 +29,17 @@ export const apiSlice = createApi({
+      query: () => "/homes",
+      transformResponse: (res) =>
+        res
+          ? res.filter((home) => home.assignment_id === null).sort(sortById)
```
You should be able to ditch the `.filter` here and just do
```
query: () => "/homes?assignment_id=null"
```
do delegate the filter to the back end!

LWest001 replied:
hmmm, this doesn't return a response for me. However, this does:
`query: () => "/homes?assignment_id"`

mattdelsordo replied:
Oh weird, it worked for me (with curl anyway)

LWest001 replied:
I tried in the app and also Postman with `localhost:3000/homes?assignment_id=null` and just get an empty array. Not sure what's going on there.

mattdelsordo commented on frontend/front/package.json:
Seems like a couple of Thiago's commits got stuck in here?

LWest001 replied:
I think this is just from rebasing to `main` after some of Thiago's fixes.

mattdelsordo commented on frontend/front/src/pages/Admin/assignment/AssignProfile.js:
```diff
@@ -5,30 +5,76 @@ import { DataGrid } from "@mui/x-data-grid";
+    isError: isAssignmentError,
+    isLoading: isAssignmentDataLoading,
+  } = useGetAssignmentQuery(aid);
+  console.log(assignmentData);
```
Remove this `console.log`

mattdelsordo commented on frontend/front/src/pages/Admin/assignment/AssignProfile.js:
```diff
@@ -81,35 +126,55 @@ const AssignProfile = () => {
+              <Button
+                variant="outlined"
+                color="primary"
+                onClick={() => navigate(`/admin/user/userprofile/1`)}
```
should this be something like
```
onClick={() => navigate(`/admin/user/userprofile/${surveyor.id}`)}
```
?

mattdelsordo commented on frontend/front/src/pages/Admin/assignment/AssignTable.js:
```diff
@@ -41,26 +45,63 @@ const AssignTable = () => {
+          : "Unassigned";
+      },
+    },
+    // { field: "home", headerName: "Home", width: 110, flex: 1 },
```
remove these comments

mattdelsordo commented on node_modules/.yarn-integrity:
Remove this

mattdelsordo commented on yarn.lock:
Remove this

mattdelsordo commented:
Haven't ran this myself but it looks pretty good from a code perspective :+1:

mattdelsordo commented on frontend/front/src/pages/Admin/assignment/AssignTable.js:
```diff
@@ -41,26 +45,61 @@ const AssignTable = () => {
+      flex: 1,
+      renderCell: (params) => {
+        let completed = 0;
+        params.row.homes.forEach((home) => {
```
You might be able to do this in a cleaner way with .reduce
